### ISSUE TITLE
Different geometry got using deepcopy

### ISSUE THREAD

**Issue by comchw** [shapely/shapely#1821]
## Expected behavior and actual behavior.

I got a GeometryCollection containing a LineString when I tried to deepcopy a GeometryCollection containing a LinearRing.

## Steps to reproduce the problem.

```
import shapely
import copy

lring = shapely.LinearRing([[0,0],[0,1],[1,1],[1,0]])
geomcollection = shapely.GeometryCollection([lring])

print(geomcollection)
print(copy.deepcopy(geomcollection))
```
Output:
```
GEOMETRYCOLLECTION (LINEARRING (0 0, 0 1, 1 1, 1 0, 0 0))
GEOMETRYCOLLECTION (LINESTRING (0 0, 0 1, 1 1, 1 0, 0 0))
```
## Operating system

Windows 11

## Shapely version and provenance

2.0.1 installed from PyPI using pip

**Comment by jorisvandenbossche** [comment]
Yes, `deepcopy` essentially pickles/unpickles the object (if the object does not define `__copy__` and `__deepcopy__`), and so what you see is a limitation of (or I would actually call it a bug in) the way we pickle geometries. 

Pickling shapely objects generally relies on WKB (de)serialization, and this has some limitations, as you list. We actually work-around the specific limitation for LinearRings by having a custom reduce method there:

https://github.com/shapely/shapely/blob/5c4670bbbac54fffddbabac202efdecdc445317d/shapely/geometry/polygon.py#L113-L116

But this only works for LinearRings itself, and not for LinearRings that are contained within a GeometryCollection.

Relying on WKT has its own set of limitations (eg loss of precision of coordinates), and is slower. So I think we certainly want to stick to WKB. 

One possible fix for GeometryCollection would be to pickle its list of children geometries, although not sure if that would not cause other corner cases.

---

Side question: what is your use case of wanting to deep copy the geometry objects? Starting with shapely 2.0, geometries are now immutable, so there generally shouldn't be a need to copy them.

**Comment by jorisvandenbossche** [comment]
Thanks. Deep copying that Feature object will then recursively copy each of its members, and thus also the geometry; that makes sense. 

In any case, also for pickling GeometryCollections in itself, I would consider this a bug, since ideally pickling always gives a faithful roundtrip (although having GeometryCollections with LinearRings is probably a rather corner case)
